[VPlan] Don't use the legacy cost model for loop conditions

john-brawn-arm opened the pull request:
The current behaviour of using the legacy cost model for instructions that compute the loop exit condition gets the wrong result when the loop has been transformed to use a different exit condition, e.g. when have tail-folded predicated vectorization the exit condition is based on the predicate vector.

Fix this by adding cost computation for BranchOnCount and removing the restriction on computing the cost for scalar ICmp/FCmp, and removing the use of the legacy cost model for loop exit conditions.

This causes quite a lot of changes to expected output in tests. Some of these are just changes to the -debug output, others are choosing a different VF due to previously over or under-estimating the cost, and in others the minimum trip count has changed as we now compute the cost for compares in the middle block.

john-brawn-arm committed:
[VPlan] Don't use the legacy cost model for loop conditions

The current behaviour of using the legacy cost model for instructions
that compute the loop exit condition gets the wrong result when we
the loop has been transformed to use a different exit condition, e.g.
when have tail-folded predicated vectorization the exit condition is
based on the predicate vector.

Fix this by adding costs for all of the VPlan instructions that can
create a compare instruction and removing the use of the legacy cost
model. This can result in the VPlan choosing a different VF to the
legacy cost model, so check for this in computeBestVF so we don't
assert.

This causes quite a lot of changes to expected output in tests. These
are mostly just changes to the -debug output, or choosing a different
VF, but others aren't as obvious as to what's happening:
 * AArch64/conditional-branches-cost.ll has more IR instructions, but
   less final machine instructions.
 * X86/pr81872.ll I've increased the trip count by 1 as otherwise the
   test no longer vectorizes.
 * X86/consecutive-ptr-uniforms.ll has more IR instructions, but the
   branches have constant condition so it would later get simplified
   to just three stores.

john-brawn-arm committed:
Do TC <= VF check differently to avoid llvm::PatternMatch error

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1117,6 +1117,29 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
+    // later gets removed.
+    Value *TC = getParent()->getPlan()->getTripCount()->getUnderlyingValue();
+    ConstantInt *TCConst = dyn_cast_if_present<ConstantInt>(TC);
+    if (TCConst && TCConst->getValue().ule(VF.getKnownMinValue()))
```
Can you add a TODO for the case where TC=vscale x M and VF=vscale * N as well? In such cases we should also be able to prove that TC <= VF because it just requires asking if M <= N.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1117,6 +1117,29 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
     return Ctx.TTI.getIndexedVectorInstrCostFromEnd(Instruction::ExtractElement,
                                                     VecTy, Ctx.CostKind, 0);
   }
+  case VPInstruction::BranchOnCount: {
```
I think BranchOnCond needs adding too.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1117,6 +1117,29 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
+    Type *ValTy = Ctx.Types.inferScalarType(getOperand(0));
+    return Ctx.TTI.getCmpSelInstrCost(Instruction::ICmp, ValTy,
+                                      CmpInst::makeCmpResultType(ValTy),
+                                      CmpInst::ICMP_EQ, Ctx.CostKind);
```
I think you need to add the cost of the branch as well.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/fully-unrolled-cost.ll:
```diff
@@ -9,11 +9,12 @@ define i64 @test(ptr %a, ptr %b) #0 {
 ; CHECK-NEXT: Cost of 0 for VF 8: induction instruction   %i.iv = phi i64 [ 0, %entry ], [ %i.iv.next, %for.body ]
-; CHECK-NEXT: Cost of 1 for VF 8: exit condition instruction   %exitcond.not = icmp eq i64 %i.iv.next, 16
 ; CHECK-NEXT: Cost of 0 for VF 8: EMIT vp<{{.+}}> = CANONICAL-INDUCTION ir<0>, vp<%index.next>
+; CHECK: Cost of 1 for VF 8: EMIT branch-on-count vp<%index.next>, vp<{{.+}}>
```
I think @igogo-x86 added this file in #106699. You might also want to remove `addFullyUnrolledInstructionsToIgnore`, since you're now dealing with the cost properly in vplan?

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/induction-costs-sve.ll:
```diff
@@ -427,71 +427,53 @@ define void @trunc_ivs_and_store(i32 %x, ptr %dst, i64 %N) #0 {
+; PRED-NEXT:    [[TMP35:%.*]] = extractelement <2 x i1> [[ACTIVE_LANE_MASK_NEXT]], i32 0
 ; PRED-NEXT:    [[TMP36:%.*]] = xor i1 [[TMP35]], true
-; PRED-NEXT:    [[VEC_IND_NEXT]] = add <4 x i32> [[VEC_IND]], splat (i32 4)
+; PRED-NEXT:    [[VEC_IND_NEXT]] = add <2 x i32> [[VEC_IND]], splat (i32 2)
```
Do you know why we're choosing a lower VF here? It may be a sensible change, but it's not obvious what triggered the different choice of VF as it doesn't seem like a FCmp, ICmp or BranchOnCount VPInstruction being generated here.

david-arm commented on llvm/test/Transforms/LoopVectorize/X86/consecutive-ptr-uniforms.ll:
```diff
@@ -174,16 +173,34 @@ attributes #0 = { "target-cpu"="knl" }
-; CHECK-NEXT:    br i1 [[CMP2]], label %[[RETURN:.*]], label %[[FOR_BODY]]
-; CHECK:       [[RETURN]]:
-; CHECK-NEXT:    ret void
+; CHECK-NEXT:  [[ENTRY:.*:]]
```
At first glance this looks worse than before.

david-arm reviewed (commented):
Thanks for moving more cost modelling into vplan. I just had a few concerns about some missing costs and a possible test regression.

fhahn reviewed (commented):
Likely related @ElvisWang123 's https://github.com/llvm/llvm-project/pull/125640. IIRC the latest status was that there still were some regressions needing investigation

john-brawn-arm committed:
Add extra comment to VPInstruction::computeCost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Use VPlanPatternMatch for counting the number of compares

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

david-arm reviewed (commented):
I'm all in favour of making steps to retire the legacy cost model where possible! However, I think there are still a few issues that need dealing with particularly as the scope of this PR is actually a bit greater than it seems.

john-brawn-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1151,6 +1151,32 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
                                                     VecTy, Ctx.CostKind, 0);
   }
+  case VPInstruction::BranchOnCount: {
+    // If TC <= VF then this is just a branch.
```
As mentioned in the comment below this, this transformation is happening in simplifyBranchConditionForVFAndUF and means the vector loop is executed exactly once. TC < VF can happen with tail folding, e.g. low_trip_count_fold_tail_scalarized_store in llvm/test/Transforms/LoopVectorize/AArch64/conditional-branches-cost.ll.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1151,6 +1151,32 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
   }
+  case VPInstruction::BranchOnCount: {
+    // If TC <= VF then this is just a branch.
+    // FIXME: Removing the branch happens in simplifyBranchConditionForVFAndUF
```
I don't think that's the way to go, we shouldn't duplicate that kind of reasoning here. The TODO should say that the branch should be simplified before we compute the costs.

As a workaround for now catching some cases here should be fine, as this should only mean we may miss some new optimizations, but not make things worse

john-brawn-arm committed:
Use getCostForRecipeWithOpcode for cmp cost

Causes a bunch of test changes, but it looks like they are due to correctly
accounting for the cost of a vector compare.
test/Transforms/LoopVectorize/X86/pr81872.ll I've adjusted the iteration count
again to that it still vectorizes.

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Fix tests after merge

david-arm reviewed (commented):
Thanks for being patient with all the comments. I spotted a bug in the icmp/fcmp cost calculation, but once that's fixed I'm happy!

john-brawn-arm committed:
Adjust cmp cost calculation

john-brawn-arm committed:
Make planContainsDifferentCompares ignore blocks outside the vector region

john-brawn-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1155,6 +1155,28 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
+  }
+  case Instruction::FCmp:
+  case Instruction::ICmp:
+    return getCostForRecipeWithOpcode(getOpcode(), VF, Ctx);
```
Done.

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update test after merge

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6988,6 +6948,35 @@ static bool planContainsAdditionalSimplifications(VPlan &Plan,
+  // Count how many compare instructions there are in the VPlan.
+  unsigned NumVPlanCompares = 0;
+  auto Iter = vp_depth_first_deep(Plan.getVectorLoopRegion()->getEntry());
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(Iter)) {
```
> planContainsDifferentCompares would count 1 compare in the legacy cost model, no compares in the vplan, and return true.

Yep, what I was wondering was if we could exclude plans with ActiveLaneMask terminated exiting blocks from the carve-out, to preserve the original check for those?

fhahn replied:
The issue with bailing out on some wide accesses should be fixed as of #169249. 

> I've changed this to ignore blocks outside of the vector loop region.
It looks like those changes may have been dropped by accident, at least it looks like the current code would still leave the parent region via the deep traversal iterator?

One example that should crash with the PR updated the current `main` and the change below is `llvm/test/Transforms/LoopVectorize/X86/float-induction-x86.ll`, but it does not, presumably because we count the compare in the middle.block?

```
diff --git a/llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp b/llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp
index c04f510dd995..048a298ef3ce 100644
--- a/llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp
+++ b/llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp
@@ -3582,6 +3582,7 @@ void VPPredInstPHIRecipe::printRecipe(raw_ostream &O, const Twine &Indent,

 InstructionCost VPWidenMemoryRecipe::computeCost(ElementCount VF,
                                                  VPCostContext &Ctx) const {
+  return 100;
   Type *Ty = toVectorTy(getLoadStoreType(&Ingredient), VF);
   unsigned AS = cast<PointerType>(Ctx.Types.inferScalarType(getAddr()))
                     ->getAddressSpace();
```

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update test after merge

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Adjust planContainsAdditionalSimplifications so loop exit conditions are counted as seen

Also need to account for TC <= VF optimization, otherwise
llvm/test/Transforms/LoopVectorize/RISCV/low-trip-count.ll fails the
assert.

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update tests

Looks like this is probably due to interaction with 4e05d702f02aea708e8431061ad886c5cc9f4f40

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/simple_early_exit.ll:
```diff
@@ -386,7 +386,7 @@ define i32 @diff_exit_block_needs_scev_check(i32 %end) {
 ; CHECK-NEXT:    [[TMP1:%.*]] = zext i10 [[TMP0]] to i64
 ; CHECK-NEXT:    [[UMAX1:%.*]] = call i64 @llvm.umax.i64(i64 [[TMP1]], i64 1)
-; CHECK-NEXT:    [[MIN_ITERS_CHECK:%.*]] = icmp ult i64 [[UMAX1]], 8
+; CHECK-NEXT:    [[MIN_ITERS_CHECK:%.*]] = icmp ult i64 [[UMAX1]], 12
```
It looks like the latest update resulted in quite a bit more changes to the minimum iteration threshold. Do you know what caused this?

john-brawn-arm replied:
Looking at the vplan, previously it had
```
    EMIT branch-on-cond vp<%11>
  No successors
}
Successor(s): middle.split

middle.split:
  EMIT vp<%13> = extract-last-part ir<%ld2>
  EMIT vp<%14> = extract-last-lane vp<%13>
  EMIT branch-on-cond vp<%9>
Successor(s): vector.early.exit, middle.block

middle.block:
  EMIT vp<%cmp.n> = icmp eq ir<64>, vp<%2>
  EMIT branch-on-cond vp<%cmp.n>
Successor(s): ir-bb<loop.end>, scalar.ph

vector.early.exit:
  EMIT vp<%17> = first-active-lane ir<%cmp>
  EMIT vp<%18> = add vp<%4>, vp<%17>
  vp<%19> = DERIVED-IV ir<3> + vp<%18> * ir<1>
Successor(s): ir-bb<loop.end>
```
now it has
```
    EMIT branch-on-two-conds vp<%9>, vp<%10>
  No successors
}
Successor(s): vector.early.exit, middle.block

middle.block:
  EMIT vp<%12> = extract-last-part ir<%ld2>
  EMIT vp<%13> = extract-last-lane vp<%12>
  EMIT vp<%cmp.n> = icmp eq ir<64>, vp<%2>
  EMIT branch-on-cond vp<%cmp.n>
Successor(s): ir-bb<loop.end>, scalar.ph

vector.early.exit:
  EMIT vp<%15> = first-active-lane ir<%cmp>
  EMIT vp<%16> = add vp<%4>, vp<%15>
  vp<%17> = DERIVED-IV ir<3> + vp<%16> * ir<1>
Successor(s): ir-bb<loop.end>
```
It looks like previously the cost of middle.block and vector.early.exit were being counted, but not middle.split. Because of 524b1788c4f5a5fd4fdda5e51dd1e484b4124870 the instructions that were in middle.split are now in middle.block, so they're being counted when previously they weren't.

So this seems correct to me? The cost that's being calculated more accurately reflects what's in the vplan.

john-brawn-arm replied:
I've applied #176754 on top of this PR at https://github.com/john-brawn-arm/llvm-project/tree/vplan_cmp_cost_pr176754. It looks like it reduces the value in the minimum iteration check in:
```
llvm/test/Transforms/LoopVectorize/AArch64/simple_early_exit.ll
llvm/test/Transforms/LoopVectorize/AArch64/sve2-histcnt-too-many-deps.ll
llvm/test/Transforms/LoopVectorize/RISCV/fminimumnum.ll
llvm/test/Transforms/LoopVectorize/RISCV/lmul.ll
llvm/test/Transforms/LoopVectorize/RISCV/strided-accesses.ll
llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-call-intrinsics.ll
llvm/test/Transforms/LoopVectorize/X86/induction-costs.ll
llvm/test/Transforms/LoopVectorize/X86/pr35432.ll
llvm/test/Transforms/LoopVectorize/X86/predicate-switch.ll
llvm/test/Transforms/LoopVectorize/vectorize-force-tail-with-evl.ll
```

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update tests after merge

john-brawn-arm commented:
Ping.

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update tests after merge

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update tests

These test changes are due to PR#176754

john-brawn-arm committed:
Update another test

fhahn reviewed (commented):
thanks for the latest update! I gave this a try on macOS, but run into an assertion for 

```
; opt -p loop-vectorize
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "arm64-apple-macosx15.0.0"

define fastcc i32 @foo(i32 %.pre215) {
entry:
  %conv.i142 = and i32 %.pre215, 3
  %add48 = add i32 %conv.i142, 3
  br label %if.end58

if.end58:                                         ; preds = %if.end58, %entry
  %indvars.iv202 = phi i64 [ 0, %entry ], [ %indvars.iv.next203, %if.end58 ]
  %ret.2188 = phi i32 [ %add48, %entry ], [ %dec67, %if.end58 ]
  %indvars.iv.next203 = add i64 %indvars.iv202, 1
  %arrayidx65 = getelementptr i8, ptr null, i64 %indvars.iv.next203
  store i8 0, ptr %arrayidx65, align 1
  %dec67 = add i32 %ret.2188, -1
  %cmp55 = icmp ugt i64 %indvars.iv202, 755
  %tobool53 = icmp eq i32 %dec67, 0
  %or.cond1 = or i1 %cmp55, %tobool53
  br i1 %or.cond1, label %if.end82.loopexit195, label %if.end58

if.end82.loopexit195:                             ; preds = %if.end58
  ret i32 0
}
```

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Distinguish between vector and scalar compares in planContainsDifferentCompares

john-brawn-arm committed:
Add test of cases where cost models disagree

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Remove exit condition check in planContainsAdditionalSimplifications

Causes the assertion to fail in some cases. Also move the call to
planContainsDifferentCompares higher in the assertion so that
the cost-model-disagree.ll test still works, as it's checking debug
output from that function.

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

fhahn reviewed (commented):
thanks for the altet update, checked the latest verison, but I think we still effectively always disable the assertion now, because the original exiting condition will get removed and then we won't find a recipe matching the compare, which so planContainsAdditionalSimplification will always return true

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Mark BranchOnCond as having the same underlying value as the cond it's replacing

Should help with stopping PlanContainsAdditionalSimplificaitons from almost
always returning true.
Has a side-effect of making BranchOnCond be affected by force-target-instruction-cost,
but previously this would have happened when pre-computing the cost of exit condition
insturctions so it works out as the same as before.

john-brawn-arm committed:
Use shallow traversal

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update unit tests

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm commented:
Ping

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
clang-format

john-brawn-arm committed:
Fix incorrect changes to tests from merge

john-brawn-arm committed:
BranchOnTwoConds uses first lane only

Fixes failure in test/Transforms/LoopVectorize/RISCV/early-exit-live-out.ll

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Remove everything to do with removing the compare in addCanonicalIVRecipes

This is no longer needed, as we only did it for
planContainsAdditionalSimplifications which no longer exists.

john-brawn-arm commented:
#190838 means that the changes to do with comparing against the legacy cost model are no longer necessary, which simplifies this to just the change to using vplan for the cmp cost.

john-brawn-arm committed:
Remove changes to SystemZ/force-target-instruction-cost.ll

These are no longer needed

john-brawn-arm commented:
I've gone through all of the test changes here to make sure they make sense. My explanation of what's going on in each:
```
Different VL due to previously underestimating the cost:
 llvm/test/Transforms/LoopVectorize/AArch64/conditional-branches-cost.ll
 llvm/test/Transforms/LoopVectorize/X86/cost-model.ll
 llvm/test/Transforms/LoopVectorize/X86/constant-fold.ll
 llvm/test/Transforms/LoopVectorize/X86/induction-costs.ll
 llvm/test/Transforms/LoopVectorize/X86/optsize.ll
 llvm/test/Transforms/LoopVectorize/X86/pr81872.ll
 llvm/test/Transforms/LoopVectorize/X86/scev-checks-unprofitable.ll
  In all these the VPlan has a vector cmp where legacy cost model was assuming scalar cmp

Different VL due to previously overestimating the cost:
 llvm/test/Transforms/LoopVectorize/AArch64/induction-costs-sve.ll
  In VPlan loop condition is based on active lane mask and cmp is removed
 llvm/test/Transforms/LoopVectorize/RISCV/low-trip-count.ll
  TC is 8, so branch-on-count has cost 0 for VL 8
 llvm/test/Transforms/LoopVectorize/X86/consecutive-ptr-uniforms.ll
  Legacy cost model has scalarized load being used in the loop condition, but VPlan has just branch-on-count
 llvm/test/Transforms/LoopVectorize/RISCV/truncate-to-minimal-bitwidth-cost.ll
  Legacy cost model has a sign-extend before the cmp which is eliminated by using branch-on-count

Higher minimum trip count due to counting cmp instructions in middle block
 llvm/test/Transforms/LoopVectorize/RISCV/conditional-scalar-assignment.ll
 llvm/test/Transforms/LoopVectorize/RISCV/inloop-reduction.ll
 llvm/test/Transforms/LoopVectorize/RISCV/partial-reduce-dot-product.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-call-intrinsics.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-cast-intrinsics.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-complex-mask.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-div.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-interleave.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-iv32.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-masked-loadstore.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-reverse-load-store.ll
 llvm/test/Transforms/LoopVectorize/RISCV/tail-folding-safe-dep-distance.ll
 llvm/test/Transforms/LoopVectorize/RISCV/transform-narrow-interleave-to-widen-memory.ll
 llvm/test/Transforms/LoopVectorize/RISCV/vectorize-vp-intrinsics.ll
 llvm/test/Transforms/LoopVectorize/X86/interleave-cost.ll

Tests that check the -debug output which has changed
 llvm/test/Transforms/LoopVectorize/AArch64/fully-unrolled-cost.ll
 llvm/test/Transforms/LoopVectorize/AArch64/scalable-vectorization-cost-tuning.ll
 llvm/test/Transforms/LoopVectorize/ARM/mve-icmpcost.ll
 llvm/test/Transforms/LoopVectorize/X86/CostModel/vpinstruction-cost.ll
 llvm/test/Transforms/LoopVectorize/X86/reduction-small-size.ll
```

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Update conditional-branches-cost.ll test

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1257,6 +1257,30 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
+                                      CmpInst::makeCmpResultType(ValTy),
+                                      CmpInst::ICMP_EQ, Ctx.CostKind);
+  }
+  case Instruction::FCmp:
```
Is it possible to add the icmp/fcmp changes in a separate PR? I think they would be useful as a standalone patch without the other bits and it might reduce the diff and make it easer to review. There are cases where we generate a vector icmp of a vector induction variable used when tail-folding. In these cases they aren't involved in the actual loop control flow at all and are merely used to predicate operations in the loop. See foo_minsize in llvm/test/Transforms/LoopVectorize/X86/optsize.ll as an example. While separately trying to remove the legacy cost model from vplan with https://github.com/llvm/llvm-project/pull/192056 I also noticed we're missing icmp/fcmp. In fact, I think the selects can be unified with the icmp/fcmp too.

In theory, you could always restrict the change to only consider VF>1 so that it doesn't interfere with the icmp used for loop control flow.

john-brawn-arm replied:
I think doing just the icmp/fcmp changes here result in double-counting the exit condition cost in some cases, but doing just the cost for vector icmp/fcmp looks like it should work. I'll create a separate PR for doing that.

david-arm replied:
OK thanks!

john-brawn-arm replied:
I've created #193268

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1286,15 +1286,30 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
-    if (vputils::onlyFirstLaneUsed(this))
-      return 0;
-    return getCostForRecipeWithOpcode(getOpcode(), VF, Ctx);
+    return getCostForRecipeWithOpcode(
```
I think in https://github.com/llvm/llvm-project/pull/193268 @fhahn mentioned that it might be possible to detect whether the scalar icmp/fcmp is being used for the exit condition. If so, I think it might be a good idea to have two PRs:

1. Adding the scalar cost model for just icmp/fcmp when the result isn't used as the exit condition.
2. A PR that adds the cost for icmp/fcmp/branchoncount when used for the exit condition.

Hopefully this reduces the risk of any one PR causing post-merge regressions and also makes it easier to review the test changes.

john-brawn-arm replied:
I've created #197146 for this.

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/partial-reduce-dot-product.ll:
```diff
@@ -61,19 +61,19 @@ define i32 @vdot4a(ptr %a, ptr %b) #0 {
+; ZVDOT4A8I-NEXT:    br i1 [[TMP9]], label [[MIDDLE_BLOCK:%.*]], label [[VECTOR_BODY]], !llvm.loop [[LOOP0:![0-9]+]]
 ; ZVDOT4A8I:       middle.block:
-; ZVDOT4A8I-NEXT:    [[TMP14:%.*]] = call i32 @llvm.vector.reduce.add.nxv1i32(<vscale x 1 x i32> [[PARTIAL_REDUCE]])
+; ZVDOT4A8I-NEXT:    [[TMP10:%.*]] = call i32 @llvm.vector.reduce.add.nxv1i32(<vscale x 1 x i32> [[PARTIAL_REDUCE]])
```
It looks like nothing changed in this file, other than renaming  and adding some missing checks. if that's the case, can you remove changes?

john-brawn-arm replied:
Changes removed

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/truncate-to-minimal-bitwidth-cost.ll:
```diff
@@ -298,27 +298,27 @@ define void @test_minbws_for_trunc(i32 %n, ptr noalias %p1, ptr noalias %p2) {
-; CHECK-NEXT:    call void @llvm.vp.scatter.nxv2i8.nxv2p0(<vscale x 2 x i8> [[TMP18]], <vscale x 2 x ptr> align 1 [[TMP19]], <vscale x 2 x i1> splat (i1 true), i32 [[TMP9]]), !alias.scope [[META12:![0-9]+]], !noalias [[META13:![0-9]+]]
-; CHECK-NEXT:    [[TMP20:%.*]] = getelementptr [1 x i64], ptr [[P2]], <vscale x 2 x i64> [[TMP12]]
-; CHECK-NEXT:    call void @llvm.vp.scatter.nxv2i64.nxv2p0(<vscale x 2 x i64> zeroinitializer, <vscale x 2 x ptr> align 8 [[TMP20]], <vscale x 2 x i1> splat (i1 true), i32 [[TMP9]]), !alias.scope [[META13]]
+; CHECK-NEXT:    [[TMP20:%.*]] = getelementptr i8, ptr [[P1]], i32 [[TMP13]]
```
just renaming changes?

john-brawn-arm replied:
Yes, I've removed these changes.

john-brawn-arm committed:
Revert a couple of tests to what they currently are on main

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1288,20 +1288,30 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
-    return getCostForRecipeWithOpcode(
-        getOpcode(), IsScalar ? ElementCount::getFixed(1) : VF, Ctx);
+    // Otherwise BranchOnCount generates ICmpEQ followed by a branch.
+    Type *ValTy = Ctx.Types.inferScalarType(getOperand(0));
```
```suggestion
    Type *ValTy = getOperand(0)->getScalarType();
```

david-arm replied:
I think this PR probably needs a rebase, since `inferScalarType` doesn't exist in HEAD>

fhahn reviewed (approved):
LGTM, thanks

john-brawn-arm committed:
Merge branch 'main' into vplan_cmp_cost

john-brawn-arm committed:
Remove use of inferScalarType
